File: javascript/momoization/allConstruct.js
const allConstruct = (target, wordBank, memo = {}) => {
    if (target in memo) return memo[target];
    if (target === '') return [[]];

    const result = [];
    
    for (let word of wordBank) {
        if (target.indexOf(word) === 0) {
            const suffix = target.slice(word.length);
            const suffixWays = allConstruct(suffix, wordBank, memo);
            console.log(suffixWays)
            const targetWays = suffixWays.map(way => [word, ...way]);
            result.push(...targetWays)
        }
    }

    memo[target] = result;
    return result;
};

// memo here doesnt improve the worst case
// time O(n^m)
// space O(m) height of recursion tree

console.log(allConstruct('skateboard', ['sk', 'ateboard', 'skate', 'board', 'asdd']))

File: ruby/memoization/allConstruct.rb

def all_construct(target, wordBank, memo={})
    return memo[target] if memo.has_key?(target)
    return [[]] if target === '';

    result = []

    wordBank.each do |word|
        if target.index(word) === 0
            suffix = target.slice(word.size, target.size - word.size)
            suffixWays = all_construct(suffix, wordBank, memo)
            targetWays = suffixWays.map do |way|
                 [word, *way]
            end
            result.push(*targetWays)
        end
        
    end
    memo[target] = result
    return result
end

p all_construct('skateboard', ['sk', 'ateboard', 'skate', 'board', 'asdd'])